app/src/main/java/org/kiwix/kiwixmobile/zim_manager/local_file_transfer/WifiDirectManager.java

```diff
@@ -0,0 +1,481 @@
+
+    manager.connect(channel, config, new WifiP2pManager.ActionListener() {
+      @Override
+      public void onSuccess() {
+        // UI updated from broadcast receiver
+      }
+
+      @Override
+      public void onFailure(int reason) {
+        String errorMessage = getErrorMessage(reason);
+        Log.d(TAG, activity.getString(R.string.connection_failed) + ": " + errorMessage);
+        displayToast(R.string.connection_failed, Toast.LENGTH_LONG);
+      }
+    });
+  }
+
+  public void performHandshakeWithSelectedPeerDevice() {
+    if (BuildConfig.DEBUG) {
+      Log.d(TAG, "Starting handshake");
+    }
+    peerGroupHandshakeAsyncTask = new PeerGroupHandshakeAsyncTask(this);
+    peerGroupHandshakeAsyncTask.execute();
+  }
+
+  public boolean isFileSender() {
+    return isFileSender;
+  }
+
+  public int getTotalFilesForTransfer() {
+    return totalFilesForTransfer;
+  }
+
+  public void setTotalFilesForTransfer(int totalFilesForTransfer) {
+    this.totalFilesForTransfer = totalFilesForTransfer;
+  }
+
+  public @NonNull ArrayList<FileItem> getFilesForTransfer() {
+    return filesForTransfer;
+  }
+
+  public void setFilesForTransfer(@NonNull ArrayList<FileItem> fileItems) {
+    this.filesForTransfer = fileItems;
+  }
+
+  public void incrementTotalFilesSent() {
+    this.filesSent++;
+  }
+
+  public @NonNull String getZimStorageRootPath() {
+    return (sharedPreferenceUtil.getPrefStorage() + "/Kiwix/");
+  }
+
+  public @NonNull InetAddress getFileReceiverDeviceAddress() {
+    return fileReceiverDeviceAddress;
+  }
+
+  public static void copyToOutputStream(@NonNull InputStream inputStream, @NonNull
+      OutputStream outputStream)
+      throws IOException {
+    byte[] bufferForBytes = new byte[1024];
+    int bytesRead;
+
+    Log.d(TAG, "Copying to OutputStream...");
+    while ((bytesRead = inputStream.read(bufferForBytes)) != -1) {
+      outputStream.write(bufferForBytes, 0, bytesRead);
+    }
+
+    outputStream.close();
+    inputStream.close();
+    Log.d(LocalFileTransferActivity.TAG, "Both streams closed");
+  }
+
+  public void setClientAddress(@Nullable InetAddress clientAddress) {
+    if (clientAddress == null) {
+      // null is returned only in case of a failed handshake
+      displayToast(R.string.device_not_cooperating, Toast.LENGTH_LONG);
+      callbacks.onFileTransferComplete();
+      return;
+    }
+
+    // If control reaches here, means handshake was successful
+    selectedPeerDeviceInetAddress = clientAddress;
+    startFileTransfer();
+  }
+
+  private void startFileTransfer() {
+    if (isGroupFormed()) {
+      if (isFileSender) {
+        Log.d(LocalFileTransferActivity.TAG, "Starting file transfer");
+
+        fileReceiverDeviceAddress =
+            (isGroupOwner()) ? selectedPeerDeviceInetAddress : getGroupOwnerAddress();
+
+        displayToast(R.string.preparing_files, Toast.LENGTH_LONG);
+        ArrayList<Uri> fileUriArrayList = new ArrayList<>();
```

Comment:
you can delete this variable

Reply:
[Done](https://github.com/kiwix/kiwix-android/pull/1276/commits/e0651dd88c2ee30b8e17d54ff215384a1d877681#diff-7e01f6f425daaca67175b74c839c6af2L342)